### PR TITLE
vm_method.c: avoid signature mismatch in rb_f_notimplement call

### PR DESCRIPTION
**Pull request by kateinoigakukun**
`rb_f_notimplement` has a similar signature with arity=-1, but it has an
extra marker argument to distinguish it from other methods in
compile-time type check in rb_define_method.
This trick is introduced to override a given arity to be -1 since
https://github.com/ruby/ruby/commit/9ef51b0b89a10c8c401cb9f2337e47a25be72cbe

However, the trailing extra argument introduces a signature mismatch
between caller and callee expectation, and it causes runtime trap on Wasm.
This patch adds rb_f_notimplement_internal, which has canonical arity=-1
signature, and makes rb_define_method family to inserts a method entry
with rb_f_notimplement_internal instead of rb_f_notimplement.

### CODE REVIEW

File: vm_method.c [hunk 1]
```diff
@@ -323,14 +323,26 @@ rb_method_table_insert(VALUE klass, struct rb_id_table *table, ID method_id, con
...
 {
     rb_notimplement();
 
     UNREACHABLE_RETURN(Qnil);
 }
 
+VALUE
+rb_f_notimplement(int argc, const VALUE *argv, VALUE obj, VALUE marker)
+{
+  rb_f_notimplement_internal(argc, argv, obj);
```
Comment (nobu) [inline vm_method.c:343]:
```suggestion
    rb_f_notimplement_internal(argc, argv, obj);
```

File: vm_method.c [hunk 2]
```diff
@@ -323,14 +323,26 @@ rb_method_table_insert(VALUE klass, struct rb_id_table *table, ID method_id, con
...
-VALUE
-rb_f_notimplement(int argc, const VALUE *argv, VALUE obj, VALUE marker)
+// rb_f_notimplement has an extra trailing argument to distinguish it from other methods
+// at compile-time to override arity to be -1. But the trailing argument introduces a
+// signature mismatch between caller and callee, so rb_define_method family inserts a
+// method entry with rb_f_notimplement_internal, which has canonical arity=-1 signature,
+// instead of rb_f_notimplement.
+NORETURN(static VALUE rb_f_notimplement_internal(int argc, const VALUE *argv, VALUE obj));
+
+static VALUE rb_f_notimplement_internal(int argc, const VALUE *argv, VALUE obj)
```
Comment (nobu) [inline vm_method.c:333]:
```suggestion
static VALUE
rb_f_notimplement_internal(int argc, const VALUE *argv, VALUE obj)
```

**Review by shyouhei — APPROVED** [review]
Ah OK.  Makes sense to me.

File: vm_method.c (hunk 1 above)
Comment (kateinoigakukun) [reply vm_method.c:343]: 🙇

**Pull request merged**